### PR TITLE
Use trust dns for hyper client resolver

### PR DESCRIPTION
**Pull request by garypen**
Investigating memory revealed that the default hyper client DNS resolver had a negative impact on the memory footprint of the router.

It may also not be respecting TTL correctly. Let's replace the default with Trust DNS.

fixes: #4030

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/Cargo.toml [hunk 1]
```diff
@@ -97,6 +97,7 @@ humantime = "2.1.0"
 humantime-serde = "1.1.1"
 hyper = { version = "0.14.27", features = ["server", "client"] }
 hyper-rustls = { version = "0.24.1", features = ["http1", "http2"] }
+hyper-trust-dns-connector = "0.4.0"
```
Comment (Geal) [inline apollo-router/Cargo.toml:100]: that crate's last release is over a year old now and it brings in old versions of trust-dns. Considering the amount of code this represent, we could have an in house implementation: https://github.com/paullgdc/hyper-trust-dns-connector/blob/master/src/lib.rs

Comment (garypen) [reply apollo-router/Cargo.toml:100]: Yep. I did have those thoughts. Ok. I'll do an in-house one since we think it's useful to have.

Comment (garypen) [reply apollo-router/Cargo.toml]: done: ef2dfcdfa9c31b9b9d07785e330be7c391ed61d0

File: apollo-router/src/services/trust_dns_connector.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    type Error = io::Error;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, name: Name) -> Self::Future {
+        let resolver = self.0.clone();
+
+        let closure = || async move {
```
Comment (Geal) [inline apollo-router/src/services/trust_dns_connector.rs:44]: why a closure here?

Comment (garypen) [reply apollo-router/src/services/trust_dns_connector.rs]: Just porting the code that was already there.

I cleaned it up: fc57d3230bba7ca9f804a80a88307e9c698847d6

**Pull request merged**